pkg/raft/raft.go

```diff
@@ -2473,6 +2473,21 @@ func (r *raft) handleSnapshot(m pb.Message) {
 	if r.restore(s) {
 		r.logger.Infof("%x [commit: %d] restored snapshot [index: %d, term: %d]",
 			r.id, r.raftLog.committed, id.index, id.term)
+		// This is safe because r.lead must be at least as up to date as m.From.
+		//
+		// Everytime we apply a snapshot, the snapshot must have come from a
+		// (once)legitimate leader that has committed everything in that snapshot.
+		//
+		// And from section 5.4 of the Raft paper, we know that:
+		// if a log entry is committed in a given term, then that entry will be
+		// present in the logs of the leaders for all higher-numbered terms.
```

Comment:
The "up-to-date" reasoning sounds vague. Suggestion:

```suggestion
		// To send MsgAppResp to any leader, we must be sure that our log is
		// consistent with that leader's log. 
		//
		// After restore(s), our log ends at the entryID of this snapshot. 
		// The snapshot came from a node at m.Term (typically the leader).
		// Everything in this snapshot has been committed during m.Term or earlier,
		// and will be present in all future term logs. It is thus safe to send
		// MsgAppResp to any leader at term >= m.Term.
		//
		// From section 5.4 of the Raft paper:
		// if a log entry is committed in a given term, then that entry will be
		// present in the logs of the leaders for all higher-numbered terms.
```